prober/derp.go

```diff
@@ -295,6 +318,188 @@ func (d *derpProber) probeBandwidth(from, to string, size int64) ProbeClass {
+	}
+
+	if err := runDerpProbeQueuingDelayContinously(ctx, from, to, fromc, toc, packetsDropped, qdh); err != nil {
+		// Record pubkeys on failed probes to aid investigation.
+		return fmt.Errorf("%s -> %s: %w",
+			fromc.SelfPublicKey().ShortString(),
+			toc.SelfPublicKey().ShortString(), err)
+	}
+	return nil
+}
+
+func runDerpProbeQueuingDelayContinously(ctx context.Context, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, packetsDropped *expvar.Float, qdh *queuingDelayHistogram) error {
+	// Circular buffer to hold packet send times. It is sized to hold timings
+	// for up to 5 seconds when sending packets at a rate of 10 per second.
+	// It assumes that packets may be dropped, but that they will generally
+	// arrive in order. Packets arriving out of order will result in older
+	// packets being ignored, effectively overcounting the number of dropped
+	// packets.
+	sentTimes := circularqueue.NewFIFO(50, func(t time.Time) {
+		// If a sent time is evicted, that means we'll never record a timing
+		// for this packet, so we considered it dropped.
+		packetsDropped.Add(1)
+	})
+
+	// Send the packets.
+	sendErrC := make(chan error, 1)
+	pkt := make([]byte, 260) // the same size as a CallMeMaybe packet observed on a Tailscale client.
+	crand.Read(pkt)
+
+	go func() {
+		t := time.NewTicker(time.Second / 10) // 10 packets per second
+		defer t.Stop()
+
+		seq := 0
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-t.C:
+				sentTimes.Push(time.Now())
+				binary.BigEndian.PutUint64(pkt, uint64(seq))
+				seq++
+				if err := fromc.Send(toc.SelfPublicKey(), pkt); err != nil {
+					sendErrC <- fmt.Errorf("sending packet %w", err)
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvFinishedC := make(chan error, 1)
+	go func() {
+		defer close(recvFinishedC) // to break out of 'select' below.
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvFinishedC <- err
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				now := time.Now()
+				if v.Source != fromc.SelfPublicKey() {
+					recvFinishedC <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				seq := binary.BigEndian.Uint64(v.Data)
+				sent := sentTimes.Pop(int(seq))
+				if sent == nil {
+					// No sent time found, ignore
+					continue
+				}
+				qdh.add(now.Sub(*sent))
+
+			case derp.KeepAliveMessage:
+				// Silently ignore.
+			default:
+				log.Printf("%v: ignoring Recv frame type %T", to.Name, v)
+				// Loop.
+			}
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("timeout: %w", ctx.Err())
+	case err := <-sendErrC:
+		return fmt.Errorf("error sending via %q: %w", from.Name, err)
+	case err := <-recvFinishedC:
+		if err != nil {
+			return fmt.Errorf("error receiving from %q: %w", to.Name, err)
+		}
+	}
+	return nil
+}
+
+// queuingDelayHistogram allows tracking a histogram of queuing delays
+type queuingDelayHistogram struct {
```

Comment:
can we not use the prometheus type natively?

Reply:
I was following the pattern that the prober framework uses for other metrics. I'm not sure of the implications of using a Histogram type directly. @awly do you have any insights?

Reply:
Looks to be related to the `ProbeClass.Metrics` func signature and label passing to "custom" metrics.

Could be refactored to make said labels available at the time the `ProbeClass` is constructed, but we're already stretching ourselves to fit into this framework, so may not justify larger refactors for this use case.

Let's update the docs to state why we're creating this histogram type on top of a prometheus histogram.

Reply:
I turned the use-specific `queuingDelayHistogram` type into a more general purpose `histogram` type and added documentation.